[export] Initial serialization

angelayi opened the pull request:
Serialization TODOs:
- [x] pytree spec: https://github.com/pytorch/pytorch/pull/102577
- [ ] higher order ops
- [ ] node metadata (specifically nn_module_stack/source_fn)
- [ ] shape env
- [ ] graph module metadata?

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #102125

angelayi committed:
[export] Initial serialization

[ghstack-poisoned]

angelayi committed:
Update on "[export] Initial serialization"

Serialization TODOs:
- [ ] pytree spec
- [ ] higher order ops
- [ ] node metadata (specifically nn_module_stack/source_fn)
- [ ] shape env
- [ ] graph module metadata?




[ghstack-poisoned]

tugsbayasgalan commented on torch/_export/serialize.py:
```diff
@@ -46,30 +69,441 @@ def _reverse_map(d):
+    return ""
+
+
+def export_signature(sig) -> GraphSignature:
```
We should probably consolidate GraphSignature with ExportMetaData fields.

angelayi replied:
Yes...this part is all tbd until ExportedProgram is figured out (cc @gmagogsfm)

tugsbayasgalan commented on torch/_export/serialize.py:
```diff
@@ -46,30 +69,441 @@ def _reverse_map(d):
+
+    def export_sym_int_op_inputs(self, args) -> List[NamedArgument]:
+        serialized_args = []
+        args_names = ["a", "b"]
```
What is "a" and "b"?

angelayi replied:
Arbitrary names because those ops don't actually have an arg name :P

tugsbayasgalan commented on torch/_export/serialize.py:
```diff
@@ -46,30 +69,441 @@ def _reverse_map(d):
+
+        buffers = {}
+        parameters = {}
+        for name, buffer in graph_module.named_buffers():
```
Can you leave a TODO here saying we will probably move this to ExportedProgram?

tugsbayasgalan commented on torch/_export/serialize.py:
```diff
@@ -46,30 +69,441 @@ def _reverse_map(d):
+
+
+        # TODO(angelayi): Graph Module metadata?
+        metadata: Dict[str, str] = {}
```
Maybe the Signature should be part of metadata.

angelayi committed:
Update on "[export] Initial serialization"

Serialization TODOs:
- [ ] pytree spec
- [ ] higher order ops
- [ ] node metadata (specifically nn_module_stack/source_fn)
- [ ] shape env
- [ ] graph module metadata?




[ghstack-poisoned]

larryliu0820 commented on torch/_export/serialize.py:
```diff
@@ -46,33 +70,440 @@ def _reverse_map(d):
+
+def export_operator(target, version) -> Operator:
+    if isinstance(target, str):
+        return Operator(name=target, version=version)
+    elif target in _SYM_INT_OPS:
+        return Operator(name=f"{target.__module__}.{target.__name__}", version=version)
+    elif isinstance(target, torch._ops.HigherOrderOperator):
+        return Operator(name=target.__name__, version=version)
+    else:
+        return Operator(name=str(target), version=version)
```
@zhxchen17  do you think we still need per-operator version after we have opset version? I think we need to make a decision on this. My hunch is that we don't need per-operator version anymore.

zhxchen17 replied:
@larryliu0820 no we don't need per op version now.

larryliu0820 commented on torch/_export/serialize.py:
```diff
@@ -46,33 +70,440 @@ def _reverse_map(d):
+        self.nodes: List[Node] = []
+        self.tensor_values: Dict[str, TensorValue] = {}
+        self.sym_int_values: Dict[str, SymInt] = {}
+        self.op_version: int = op_version
```
oh interesting, so this is actually an operator set version but not per-operator version. Then this should be the constant along the lifecycle of `Serializer` then. I would suggest move this out from `Serializer` for now. In the end it should come from the version table.

zhxchen17 replied:
agree

angelayi committed:
Update on "[export] Initial serialization"

Serialization TODOs:
- [ ] pytree spec
- [ ] higher order ops
- [ ] node metadata (specifically nn_module_stack/source_fn)
- [ ] shape env
- [ ] graph module metadata?




[ghstack-poisoned]

angelayi committed:
Update on "[export] Initial serialization"

Serialization TODOs:
- [ ] pytree spec
- [ ] higher order ops
- [ ] node metadata (specifically nn_module_stack/source_fn)
- [ ] shape env
- [ ] graph module metadata?




[ghstack-poisoned]

angelayi committed:
Update on "[export] Initial serialization"

Serialization TODOs:
- [ ] pytree spec
- [ ] higher order ops
- [ ] node metadata (specifically nn_module_stack/source_fn)
- [ ] shape env
- [ ] graph module metadata?




[ghstack-poisoned]

angelayi committed:
Update on "[export] Initial serialization"

Serialization TODOs:
- [ ] pytree spec
- [ ] higher order ops
- [ ] node metadata (specifically nn_module_stack/source_fn)
- [ ] shape env
- [ ] graph module metadata?




[ghstack-poisoned]

angelayi committed:
Update on "[export] Initial serialization"

Serialization TODOs:
- [ ] pytree spec
- [ ] higher order ops
- [ ] node metadata (specifically nn_module_stack/source_fn)
- [ ] shape env
- [ ] graph module metadata?




[ghstack-poisoned]

zhxchen17 commented on torch/_export/serialize.py:
```diff
@@ -1,21 +1,44 @@
-from typing import Optional, Tuple
+import io
```
could you move the entire file under `serde/`

zhxchen17 commented on torch/_export/serialize.py:
```diff
@@ -46,33 +69,411 @@ def _reverse_map(d):
+        )
+
+
+def serialize(exported_program: ep.ExportedProgram) -> Tuple[GraphModule, bytes]:
```
Q1: Should we pass in a version table here?

Q2: Shouldn't we only return a bytes object here?

zhxchen17 commented on torch/_export/serialize.py:
```diff
@@ -46,33 +69,411 @@ def _reverse_map(d):
+        if node.target is operator.getitem:
+            return
+
+        if node.target is torch.set_grad_enabled:
+            # Hack for torch.no_grad support. In the long run this should become
+            # a higher order op but this is fine for now. See [NOTE: nograd support]
+            ex_node = Node(
+                target=export_operator("torch.set_grad_enabled"),
+                inputs=[NamedArgument(name="arg", arg=self.export_input(node.args[0]))],
+                outputs=[],
+                metadata=export_metadata(node),
+            )
```
Could we exclude this?

zhxchen17 commented on torch/_export/serialize.py:
```diff
@@ -46,33 +69,411 @@ def _reverse_map(d):
+    operator.mod,
+}
+
+def export_sym_int(s: Union[int, torch.SymInt]) -> SymInt:
```
def serialize_sym_int ?

zhxchen17 commented on torch/_export/serialize.py:
```diff
@@ -46,33 +69,411 @@ def _reverse_map(d):
 
 
-def _extract_tensor_meta(result: torch.Tensor) -> TensorMeta:
+def export_tensor_meta(t: torch.Tensor) -> TensorMeta:
```
def serialize_tensor_meta?

zhxchen17 commented on torch/_export/serialize.py:
```diff
@@ -46,33 +69,411 @@ def _reverse_map(d):
     )
 
 
+def export_metadata(node: torch.fx.Node) -> Dict[str, str]:
```
def serialize_node_meta?

zhxchen17 commented on torch/_export/serialize.py:
```diff
@@ -46,33 +69,411 @@ def _reverse_map(d):
+    return ret
+
+
+def export_operator(target) -> str:
```
def serialize_operator?

zhxchen17 commented on torch/_export/serialize.py:
```diff
@@ -46,33 +69,411 @@ def _reverse_map(d):
+        return str(target)
+
+
+def export_call_spec(call_spec: ep.CallSpec) -> CallSpec:
```
same

zhxchen17 commented on torch/_export/serialize.py:
```diff
@@ -46,33 +69,411 @@ def _reverse_map(d):
+    return CallSpec(in_spec="", out_spec="")
+
+
+def export_signature(sig: ep.ExportGraphSignature) -> GraphSignature:
```
same

zhxchen17 commented on torch/_export/serialize.py:
```diff
@@ -46,33 +69,411 @@ def _reverse_map(d):
+    return graph_signature
+
+
+def export_state_dict(state_dict: Dict[str, Any]) -> bytes:
```
same

zhxchen17 commented on torch/_export/serialize.py:
```diff
@@ -46,33 +69,411 @@ def _reverse_map(d):
+            isinstance(arg, torch.fx.Node) and arg.name in self.sym_int_values
+        )
+
+    def export_input(self, arg) -> Argument:
```
def serialize_input

zhxchen17 commented on torch/_export/serialize.py:
```diff
@@ -46,33 +69,411 @@ def _reverse_map(d):
+###############################################################################
+
+
 def convert_fake_tensor_to_tensor_meta(
```
do we still need this?

zhxchen17 commented on torch/_export/serialize.py:
```diff
@@ -102,14 +503,14 @@ def get_shape_env(val) -> Optional[ShapeEnv]:
 
 
-def convert_tensor_meta_to_fake_tensor(ep: ExportedProgram, shape_env: ShapeEnv = None) -> ExportedProgram:
+def convert_tensor_meta_to_fake_tensor(ep: ep.ExportedProgram, shape_env: ShapeEnv = None) -> ep.ExportedProgram:
```
same, do we still need this?

zhxchen17 reviewed (commented):
looks great, lots of questions tho.

larryliu0820 commented on torch/_export/serialize.py:
```diff
@@ -46,33 +69,411 @@ def _reverse_map(d):
+        self.nodes: List[Node] = []
+        self.tensor_values: Dict[str, TensorValue] = {}
+        self.sym_int_values: Dict[str, SymInt] = {}
+        self.opset_version: int = opset_version
```
Is it possible to do a `Dict[str, Int]` for opset_version? I think it's clear that we want to serialize an opset version per namespace. cc @avikchaudhuri

zhxchen17 replied:
yeah I think we should pass in a dict.

angelayi committed:
Update on "[export] Initial serialization"

Serialization TODOs:
- [ ] pytree spec: https://github.com/pytorch/pytorch/pull/102577
- [ ] higher order ops
- [ ] node metadata (specifically nn_module_stack/source_fn)
- [ ] shape env
- [ ] graph module metadata?




[ghstack-poisoned]

avikchaudhuri commented on torch/_export/serde/schema.py:
```diff
@@ -154,15 +148,23 @@ class GraphSignature:
+class ExportedProgram:
+    graph_module: GraphModule
     signature: GraphSignature
+    call_spec: CallSpec
```
Can you describe what the purpose of `call_spec` is? Won't `graph_module` have the `call_spec` mappings burned in?

Is there a standard transform that uses this information?

angelayi replied:
call_spec contains the in_spec/out_spec which is used when we're running the graph module eagerly so that the inputs/output formats match how we would run the original function eagerly.

avikchaudhuri replied:
Bit confused about why we also then have flatten / unflatten calls in a graph module...

angelayi replied:
Mm, it defines the way we pass inputs and receive outputs to the graph module. Zhengxu stated below that it makes more sense to put the GraphSignature in GraphModule, and I think CallSpec should go with wherever the Signature goes.

avikchaudhuri reviewed (approved):
Accepting to unblock, lgtm. See comments though.

angelayi commented on torch/_export/serde/schema.py:
```diff
@@ -112,15 +112,9 @@ class NamedArgument:
 @dataclass
 class Node:
-    target: Operator
+    target: str
```
yup, it'll reside in the toplevel GraphModule

angelayi committed:
Update on "[export] Initial serialization"

Serialization TODOs:
- [ ] pytree spec: https://github.com/pytorch/pytorch/pull/102577
- [ ] higher order ops
- [ ] node metadata (specifically nn_module_stack/source_fn)
- [ ] shape env
- [ ] graph module metadata?




[ghstack-poisoned]

larryliu0820 commented on torch/_export/serde/schema.py:
```diff
@@ -154,15 +148,23 @@ class GraphSignature:
-    buffers: Dict[str, TensorMeta]
-    parameters: Dict[str, TensorMeta]
-    metadata: Dict[str, str]
+    opset_version: Dict[str, int]
```
My naive approach is to serials a dict directly. It will be a namespace to op set version mapping. Example: {“aten”: 3, “custom_namespace”: 4}. We can change this in the future if needed.

avikchaudhuri replied:
OK, so in any case, no mapping of op names (also str) to versions, yeah? That mapping will exist somewhere else in a table / code?

angelayi commented on torch/_export/serde/schema.py:
```diff
@@ -143,7 +137,7 @@ class Graph:
 class BackwardSignature:
     gradients_to_parameters: Dict[str, str]
-    gradients_to_userInputs: Dict[str, str]
+    gradients_to_user_inputs: Dict[str, str]
```
not sure..copied from the internal thrift file

zhxchen17 commented on torch/_export/serde/schema.py:
```diff
@@ -154,15 +148,23 @@ class GraphSignature:
-    buffers: Dict[str, TensorMeta]
-    parameters: Dict[str, TensorMeta]
-    metadata: Dict[str, str]
+    opset_version: Dict[str, int]
```
@angelayi I think it makes more sense to me if we put opset_version in ExportedProgram, and put signature in GraphModule. The rationale is to closely mirror the schema in sigmoid.
